forc-pkg/src/pkg.rs

```diff
@@ -2971,6 +2985,67 @@ pub fn build(
         let mut source_map = SourceMap::new();
         let pkg = &plan.graph()[node];
         let manifest = &plan.manifest_map()[&pkg.id()];
+        let is_contract_dependency = is_contract_dependency(plan.graph(), node);
+        // If we are building a contract and tests are enabled or we are building a contract
+        // dependency, we need the tests exlcuded version.
+        let contract_without_tests = if (include_tests
+            && matches!(manifest.program_type(), Ok(TreeType::Contract)))
+            || is_contract_dependency
+        {
+            // We will build a contract with tests enabled, we will also need the same contract with tests
+            // disabled for:
+            //
+            //   1. Interpreter deployment in `forc-test`.
+            //   2. Contract ID injection in `forc-pkg` if this is a contract dependency to any
+            //      other pkg, so that injected contract id is not effected by the tests.
+            //
+            let dep_namespace = match dependency_namespace(
+                &lib_namespace_map,
+                &compiled_contract_deps,
+                &plan.graph,
+                node,
+                manifest.config_time_constants(),
+                engines,
+            ) {
+                Ok(o) => o,
+                Err(errs) => {
+                    print_on_failure(profile.terse, &[], &errs);
+                    bail!("Failed to compile {}", pkg.name);
+                }
+            };
+            let profile = BuildProfile {
+                include_tests: false,
+                ..profile.clone()
+            };
+            let (built_package_without_tests, _) = compile(
+                pkg,
+                manifest,
+                target,
+                &profile,
+                dep_namespace,
+                engines,
+                &mut source_map,
+            )?;
+            // If this contract is built because tests are enabled we need to insert CONTRACT_ID
+            // for the contract.
+            if !is_contract_dependency {
+                // `forc-test` interpreter deployments are done with zeroed salt.
+                let contract_id =
+                    contract_id(&built_package_without_tests, &fuel_tx::Salt::zeroed());
+                let contract_id_constant_name = CONTRACT_ID_CONSTANT_NAME.to_string();
+                let contract_id_value = format!("0x{contract_id}");
+                let contract_id_constant = ConfigTimeConstant {
+                    r#type: "b256".to_string(),
+                    value: contract_id_value.clone(),
+                    public: true,
+                };
+                let constant_declarations = vec![(contract_id_constant_name, contract_id_constant)];
+                const_inject_map.insert(pkg.clone(), constant_declarations);
+            }
+            Some(built_package_without_tests)
+        } else {
+            None
+        };
```

Comment:
This section copies a lot of the same logic from below already used to construct a `BuiltPackage`. Can we abstract this out into a function and share some of the logic with the section below?

Reply:
Also, it looks like this only builds *contracts* without tests, but I suspect we want to build *all* non-member dependencies without tests (as mentioned [here](https://github.com/FuelLabs/sway/pull/4159#pullrequestreview-1310438754)).